File: crutches/crutches.cpp
#define _WIN32_WINNT 0x501 // XP
#define _WIN32_IE 0x501

#include "crutches.h"

#include <queue>

#include <commctrl.h>

using namespace std;

#define WM_RESIZE_KEY 0
#define WM_COMMAND_KEY 1
#define WM_CONTEXTMENU_KEY 2
#define WM_ITEMCHANGED_KEY 3
#define WM_ITEMACTIVATE_KEY 4
#define WM_CLOSE_KEY 5

UINT _resizeMsgId;
UINT _commandMsgId;
UINT _contextMenuMsgId;
UINT _itemChangedMsgId;
UINT _itemActivateMsgId;
UINT _closeMsgId;

queue<Message> msgQueue;


UINT WINAPI GetRegisteredMessageId(UINT key)
{
    switch (key) {
    case WM_RESIZE_KEY:
        if (0 == _resizeMsgId) {
            _resizeMsgId = RegisterWindowMessageW(L"resize_0b0f95e6-7ef7-4767-b484-940e7a3cf4f1");
        }
        return _resizeMsgId;

    case WM_COMMAND_KEY:
        if (0 == _commandMsgId) {
            _commandMsgId = RegisterWindowMessageW(L"command_442946bf-f806-434b-baa3-98439930eecd");
        }
        return _commandMsgId;

    case WM_CONTEXTMENU_KEY:
        if (0 == _contextMenuMsgId) {
            _contextMenuMsgId = RegisterWindowMessageW(L"contextmenu_50fe6189-a94b-4826-8dc3-48e179f89ffc");
        }
        return _contextMenuMsgId;

    case WM_ITEMCHANGED_KEY:
        if (0 == _itemChangedMsgId) {
            _itemChangedMsgId = RegisterWindowMessageW(L"itemchanged_b453604c-2195-4df2-8d9d-e6486d9bf73a");
        }
        return _itemChangedMsgId;

    case WM_ITEMACTIVATE_KEY:
        if (0 == _itemActivateMsgId) {
            _itemActivateMsgId = RegisterWindowMessageW(L"itemactivate_1ff744a7-ff21-464c-b0a7-786843b75976");
        }
        return _itemActivateMsgId;

    case WM_CLOSE_KEY:
        if (0 == _closeMsgId) {
            _closeMsgId = RegisterWindowMessageW(L"close_cd8f8d08-cdb1-42c0-9d8c-bcf63f4114d7");
        }
        return _closeMsgId;
    }

    return 0;
}

int WINAPI GetCustomMessage(Message* msg)
{
    if (0 == msg) {
        SetLastError(ERROR_INVALID_PARAMETER);
        return -1;
    }

    if (msgQueue.empty()) {
        return 0;
    }

    (*msg) = msgQueue.back();
    msgQueue.pop();

    return msgQueue.size() + 1;
}

bool EnqueueMessage(const Message &msg)
{
    if (msgQueue.size() >= 1000) {
        return false;
    }

    msgQueue.push(msg);

    return true;
}

LRESULT CALLBACK ContainerWndProc(HWND hwnd, UINT msg, WPARAM wParam, LPARAM lParam)
{
    switch(msg)
    {
    case WM_CLOSE:
        PostMessageW(hwnd, _closeMsgId, wParam, lParam);
        break;

    case WM_CONTEXTMENU:
        PostMessageW(hwnd, _contextMenuMsgId, wParam, lParam);

    case WM_COMMAND:        
        PostMessageW(hwnd, _commandMsgId, wParam, lParam);
        break;

    case WM_NOTIFY:
        switch (((NMHDR*)(lParam))->code) {
        case LVN_ITEMCHANGED: {
                NMLISTVIEW* nmlv = (NMLISTVIEW*)lParam;

                Message m;
                m.hwnd = nmlv->hdr.hwndFrom;
                m.msg = _itemChangedMsgId;
                m.wParam = 0;
                m.lParam = nmlv->iItem;

                EnqueueMessage(m);
                break;
            }

        case LVN_ITEMACTIVATE: {
                NMITEMACTIVATE* nmia = (NMITEMACTIVATE*)lParam;

                Message m;
                m.hwnd = nmia->hdr.hwndFrom;
                m.msg = _itemActivateMsgId;
                m.wParam = 0;
                m.lParam = nmia->iItem;

                EnqueueMessage(m);
                break;
            }
        }

        break;

    case WM_SIZE:
    case WM_SIZING: {
            LRESULT ret = DefDlgProcW(hwnd, msg, wParam, lParam);
            PostMessageW(hwnd, _resizeMsgId, 0, 0);
            return ret;
        }

    default:
        return DefDlgProcW(hwnd, msg, wParam, lParam);
    }
    return 0;
}

ATOM WINAPI RegisterWindowClass(HINSTANCE hInst)
{
    WNDCLASSEX wcButton;
    WNDCLASSEX wc;

    wc.hCursor       = LoadCursor(hInst, IDC_ARROW);

    if (GetClassInfoExW(hInst, L"BUTTON", &wcButton) != FALSE) {
        wc.hCursor = wcButton.hCursor;
    }

    wc.cbSize        = sizeof(WNDCLASSEX);
    wc.style         = 0;
    wc.lpfnWndProc   = ContainerWndProc;
    wc.cbClsExtra    = 0;
    wc.cbWndExtra    = DLGWINDOWEXTRA;
    wc.hInstance     = hInst;
    wc.hIcon         = LoadIconW(hInst, IDI_APPLICATION);
//    wc.hCursor       = LoadCursor(hInst, IDC_ARROW);
    wc.hbrBackground = (HBRUSH)(COLOR_3DFACE+1);
    wc.lpszMenuName  = 0;
    wc.lpszClassName = L"Container_WindowClass";
    wc.hIconSm       = LoadIcon(hInst, IDI_APPLICATION);

    return RegisterClassExW(&wc);
}

File: crutches/crutches.go
// Copyright 2010 The Walk Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package crutches

import (
	"unsafe"
)

import (
	. "walk/winapi"
	. "walk/winapi/kernel32"
	. "walk/winapi/user32"
)

// Registered message keys
const (
	_WM_RESIZE_KEY       = 0
	_WM_COMMAND_KEY      = 1
	_WM_CONTEXTMENU_KEY  = 2
	_WM_ITEMCHANGED_KEY  = 3
	_WM_ITEMACTIVATE_KEY = 4
	_WM_CLOSE_KEY        = 5
)

// Functions
var (
	_getRegisteredMessageId uint32
	_registerWindowMessage  uint32
)

// Registered message ids
var (
	closeMsgId        uint
	commandMsgId      uint
	contextMenuMsgId  uint
	itemActivateMsgId uint
	itemChangedMsgId  uint
	resizeMsgId       uint
)

func CloseMsgId() uint {
	return closeMsgId
}

func CommandMsgId() uint {
	return commandMsgId
}

func ContextMenuMsgId() uint {
	return contextMenuMsgId
}

func ItemActivateMsgId() uint {
	return itemActivateMsgId
}

func ItemChangedMsgId() uint {
	return itemChangedMsgId
}

func ResizeMsgId() uint {
	return resizeMsgId
}

type Message struct {
	Hwnd   HWND
	Msg    uint
	WParam uintptr
	LParam uintptr
}

// Internal functions defined in .c files
func initqueue()
func initcrutch()
func getCustomMessage(msgPointer uintptr) uintptr
func registerWindowClass(hInstance uintptr) uintptr
func getRegisteredMessageId(key uint) uint

func init() {
	initcrutch()
	initqueue()

	// Library
	user32 := MustLoadLibrary("user32.dll")

	// Functions
	_registerWindowMessage = MustGetProcAddress(user32, "RegisterWindowMessageW")

	resizeMsgId = getRegisteredMessageId(_WM_RESIZE_KEY)
	// resizeMsgId = getRegisteredMessage2("resize_0b0f95e6-7ef7-4767-b484-940e7a3cf4f1")
	commandMsgId = getRegisteredMessageId(_WM_COMMAND_KEY)
	contextMenuMsgId = getRegisteredMessageId(_WM_CONTEXTMENU_KEY)
	itemChangedMsgId = getRegisteredMessageId(_WM_ITEMCHANGED_KEY)
	itemActivateMsgId = getRegisteredMessageId(_WM_ITEMACTIVATE_KEY)
	closeMsgId = getRegisteredMessageId(_WM_CLOSE_KEY)
}

func GetCustomMessage(msg *Message) int {
	ret := getCustomMessage(uintptr(unsafe.Pointer(msg)))
	return int(ret)
}
/*
func getRegisteredMessage2(msgid string) uint {
    ret, _, _ := syscall.Syscall(uintptr(_registerWindowMessage),
        uintptr(unsafe.Pointer(syscall.StringToUTF16Ptr(msgid))),
        0,
        0)
    return uint(ret)
}
*/

func RegisterWindowClass(hInstance HINSTANCE) ATOM {
	ret := registerWindowClass(uintptr(hInstance))
	return ATOM(ret)
}
